python/ray/experimental/channel/serialization_context.py

```diff
@@ -15,21 +15,39 @@ def __init__(self):
         # execute one task at a time in `do_exec_tasks`. It will not execute multiple
         # Ray tasks on a single actor simultaneously.
         self.intra_process_channel_buffers: Dict[str, Any] = {}
+        # The number of readers for each channel. When the number of readers
+        # reaches 0, remove the data from the buffer.
+        self.channel_id_to_num_readers: Dict[str, int] = {}
 
     def set_use_external_transport(self, use_external_transport: bool) -> None:
         self.use_external_transport = use_external_transport
 
-    def set_data(self, channel_id: str, value: Any) -> None:
+    def set_data(self, channel_id: str, value: Any, num_readers: int) -> None:
+        assert num_readers > 0, "num_readers must be greater than 0."
```

Comment:
A newbie question here: Seems that there will be one `_SerializationContext` per DAG actor. If there's nobody reading the returned value of this node (this is a leaf node), will it raise an error?

Reply:
> Seems that there will be one _SerializationContext per DAG actor. 

You are correct.

> If there's nobody reading the returned value of this node (this is a leaf node), will it raise an error?

For IntraProcessChannel, if there is no reader, the channel will not be created. You can see the logic in `shared_memory_channel.py`.

```python
        if num_local_readers > 0:
            local_channel = IntraProcessChannel(num_local_readers)
```

Reply:
For the leaf node, I think we currently don't do anything. Maybe we should raise a ValueError or warning message.

Reply:
i see. from my experiments, these leaf node will not be executed.

It's not a blocker for now, since there will be no leaf node in PP. But for FSDP, there might be some collective calls which doesn't has return values, thus becomes leaf nodes.